src/gui/src/PatternEditor/PatternEditorPanel.cpp

```diff
@@ -884,6 +870,85 @@ void PatternEditorPanel::patternSizeChanged( int nSelected )
 	EventQueue::get_instance()->push_event( EVENT_SELECTED_PATTERN_CHANGED, -1 );
 }
 
+void PatternEditorPanel::updatePatternSizeLCD(){
+	// update pattern size
+	int nPatternSize = m_pPattern->get_length();
+	int den = m_pPattern->get_denominator();
+	int nBeatUnit = MAX_NOTES / den;
+
+	char tmp[20];
+	if (nPatternSize % nBeatUnit == 0) {
+		sprintf( tmp, "%d/%d", nPatternSize / nBeatUnit, den);
+	} 
+	else {
+		sprintf( tmp, "%.3f/%d", (float) nPatternSize / nBeatUnit, den);
+	}
+	__pattern_size_LCD->setText( tmp );//
+}
+
+void PatternEditorPanel::patternSizeLCDClicked()
+{
+	Hydrogen *pEngine = Hydrogen::get_instance();
+	if ( pEngine->getState() != STATE_READY ) {	
+		QMessageBox::information( this, "Hydrogen", tr( "Is not possible to change the pattern size when playing." ) );
+		return;
+	} //TODO is it really impossible to change the pattern size when playing?
+
+	bool bIsOkPressed;
+	int denominator;
+	
+	QString qtmp = QInputDialog::getText( this, "Hydrogen",
+				tr( "New Pattern Size\nIndicate numerator / denominator (eg. \"4/4\") or just numerator"),
+				QLineEdit::Normal, __pattern_size_LCD->getText(), &bIsOkPressed );
+	
+	if ( bIsOkPressed  ) {		
+		QStringList parts = qtmp.split('/');
+		int nDenominator = m_pPattern->get_denominator();
+		if ( parts.size() == 1 || parts.size() == 2) { // must reject if parts.size > 2 or null
+		    bool bOk;
+		    double fNumerator = parts[0].toDouble( &bOk );
+		    if ( bOk && parts.size() == 2 ) {
+			nDenominator = parts[1].toInt( &bOk );
+			if (bOk && (nDenominator <= 0 || nDenominator > 192) ){
+			   QMessageBox::information( this, "Hydrogen", tr( "Denominator value rejected.\nLimits: (0, 192]" ) );
+			   return;
+			}
+		    }
+		    if ( bOk && fNumerator > 0) {
+			if (fNumerator / nDenominator > 4.){ 
+			     //this is limited because the pattern editor ruler goes up to 16/4. Limit might be extended
+				QMessageBox::information( this, "Hydrogen", tr( "Pattern size too big.\nMaximum = 16/4" ) );
+				return;
+			}
+			else {	 
+				if ( MAX_NOTES % nDenominator != 0 ){
+					QMessageBox::information( this, "Hydrogen",
+						tr("Warning: since finite resolution, Hydrogen can handle precisely only "
+						"denominators that divide 192 (i.e. powers of 2, also multiplied by 3)") );
+				}
+				// set numerator and denominator
+				double fLength = MAX_NOTES / (double) nDenominator * fNumerator;
+				if (fLength - round(fLength) != 0){
+					fLength = round(fLength);
+					QMessageBox::information( this, "Hydrogen",
+						tr("Pattern size was rounded.\n(resolution = 48 ticks/quarter note)" ));
+				}
+				m_pPattern->set_length( (int) fLength);
+				m_pPattern->set_denominator( nDenominator );
+				patternLengthChanged();
```

Comment:
I'm not sure if it's intentional but this accepts dominators that aren't factors of `MAX_NOTES`. I don't think those should be accepted because I think that helps users *understand* what's going on (and get an intuitive feel for it) more than explanatory text does, and because the approximation misrepresents the capabilities the program has.

Thinking as a user, if I were aiming for some specific timing, I wouldn't want the program to accept something that's not *quite* right (with an error message that I didn't take the time to read). I'd rather it reject that value and show me what *is* valid so that I can find some other way of achieving what I want (eg. changing the BPM).

Does that make sense?

Reply:
Personally I prefer to have the more possibilities even if they are not precise... hence the warnings and not rejections for that values

Reply:
Another thought: if we allow to set decimal values with respect to quarter notes, we are allowing any pattern size (in ticks) that we can achieve with any other denominator.
For example both the entries 0.79/4 and 1/5 generate the same pattern size of 38 ticks. So I think there's no need of rejecting denominators that are not divisors of MAX_NOTES, nor I would like to quantize more the entry as the decimal entry seems a powerful feature to me.

Reply:
> The user would see that he cannot achieve 1/5 because the LCD will show 0.990/5

I like this reasoning, and I think I agree. My only concern would be, will they be able to work out *why* they can't get an exact size?

So a suggestion (perhaps for later, rather than having it block this PR): how about using tooltips and a warning icon (cc. #903) next to the LCD as a hint for the user?

Reply:
Yes I agree

Reply:
The warning button is ready.
Can you suggest me the text for the tooltip and for the opening dialog after button is clicked?

Reply:
This is feeling pretty good! :)

Text suggestions.

For the immediate warning message on entering an unsupported denominator, I think it's probably good to make this short, with more detail in the warning icon click through.

So initially:

>  "Pattern length in 1/%1 notes is not supported. Length will be approximated."

(passing the user-entered denominator in arg())

Tooltip text suggestion:

>  "Unsupported note denominator. Click for more information."

and then for the information dialog:

>  "Hydrogen can only represent notes as small as 1/%1 of a whole note, so note denominators must be factors of this.
> Supported values are: 1/1, 1/2, 1/3, 1/4, 1/6, 1/8, 1/12, 1/16, 1/24, 1/32, 1/48, 1/64, 1/96, 1/192"

Any thoughts? :)

Reply:
> "Pattern length in 1/%1 notes is not supported. Length will be approximated."

Ok, even if uselessly unconventional entries like 10/5 will not be approximated...

> "Unsupported note denominator. Click for more information."

Perfect!

>"Hydrogen can only represent notes as small as 1/%1 of a whole note, so note denominators must be factors of _this_.
Supported values are: 1/1, 1/2, 1/3, 1/4, 1/6, 1/8, 1/12, 1/16, 1/24, 1/32, 1/48, 1/64, 1/96, 1/192"

 _this_ is intended MAX_NOTES,  but may be confused with 1/MAX_NOTES...

Reply:
Similar to what I was thinking before: should the warning icon appear even if `nPatternSize % nDen == 0 `?
For example: 5/5 (even if useless) in this case the length (192 ticks) is shown correctly with such denominator of course.
In your opinion

Reply:
Another thought: should the warning icon appear even if `( nPatternSize * nDen ) % MAX_NOTES == 0`?
For example a useless 5/5 (192 ticks) that is shown correctly with such denominator

Reply:
> Good point! In which case, maybe better to make "will be approximated" into "may be approximated"

Ok

> Maybe expressing it in terms of "note values must be multiples of..."

Yes, I would prefer this

> I think either way would be fine. The only risk I can think of in skipping the warning in this case would be that the user might write something in, say, 5/5 and then decide to add a pattern that's a beat longer and then find out at that late stage that 6/5 doesn't really work.

I agree, let's keep the warning